les/utils/limiter.go

```diff
@@ -0,0 +1,405 @@
+// there before.
+func (ag *addressGroup) add(nq *nodeQueue) {
+	if nq.groupIndex != -1 {
+		panic("added node queue is already in an address group")
+	}
+	l := len(ag.nodes)
+	if l == 1 {
+		ag.nodeSelect = NewWeightedRandomSelect(flatWeight)
+		ag.nodeSelect.Update(ag.nodes[0])
+	}
+	nq.groupIndex = l
+	ag.nodes = append(ag.nodes, nq)
+	ag.sumFlatWeight += nq.flatWeight
+	ag.groupWeight = ag.sumFlatWeight / uint64(l+1)
+	if l >= 1 {
+		ag.nodeSelect.Update(ag.nodes[l])
+	}
+}
+
+// update updates the selection weight of the node queue inside the address group.
+// It is the caller's responsibility to update the group's selection weight in the
+// address selector.
+func (ag *addressGroup) update(nq *nodeQueue, weight uint64) {
+	if nq.groupIndex == -1 || nq.groupIndex >= len(ag.nodes) || ag.nodes[nq.groupIndex] != nq {
+		panic("updated node queue is not in this address group")
+	}
+	ag.sumFlatWeight += weight - nq.flatWeight
+	nq.flatWeight = weight
+	ag.groupWeight = ag.sumFlatWeight / uint64(len(ag.nodes))
+	if ag.nodeSelect != nil {
+		ag.nodeSelect.Update(nq)
+	}
+}
+
+// remove removes the node queue from the address group. It is the caller's responsibility
+// to remove the address group from the address map if it is empty.
+func (ag *addressGroup) remove(nq *nodeQueue) {
+	if nq.groupIndex == -1 || nq.groupIndex >= len(ag.nodes) || ag.nodes[nq.groupIndex] != nq {
+		panic("removed node queue is not in this address group")
+	}
+
+	l := len(ag.nodes) - 1
+	if nq.groupIndex != l {
+		ag.nodes[nq.groupIndex] = ag.nodes[l]
+		ag.nodes[nq.groupIndex].groupIndex = nq.groupIndex
+	}
+	nq.groupIndex = -1
+	ag.nodes = ag.nodes[:l]
+	ag.sumFlatWeight -= nq.flatWeight
+	if l >= 1 {
+		ag.groupWeight = ag.sumFlatWeight / uint64(l)
+		if l == 1 {
+			ag.nodeSelect = nil
+		} else {
+			ag.nodeSelect.Remove(nq)
+		}
+	} else {
+		ag.groupWeight = 0
+	}
+}
+
+// choose selects one of the node queues belonging to the address group
+func (ag *addressGroup) choose() *nodeQueue {
+	if ag.nodeSelect == nil { // nodes list should never be empty here
+		return ag.nodes[0]
+	}
+	return ag.nodeSelect.Choose().(*nodeQueue)
+}
+
+// NewLimiter creates a new Limiter
+func NewLimiter(totalPriorLimit uint64) *Limiter {
+	l := &Limiter{
+		addressSelect:   NewWeightedRandomSelect(func(item interface{}) uint64 { return item.(*addressGroup).groupWeight }),
+		valueSelect:     NewWeightedRandomSelect(func(item interface{}) uint64 { return item.(*nodeQueue).valueWeight }),
+		nodes:           make(map[enode.ID]*nodeQueue),
+		addresses:       make(map[string]*addressGroup),
+		totalPriorLimit: totalPriorLimit,
+	}
+	l.cond = sync.NewCond(&l.lock)
+	go l.processLoop()
+	return l
+}
+
+// selectionWeights calculates the selection weights of a node for both the address and
+// the value selector. The weight depends on the relative cost of the previous processed
+// request.
+func selectionWeights(relCost, value float64) (flatWeight, valueWeight uint64) {
```

Comment:
As I commented, the flatweight is derived by some means. It's not fair anymore...

Reply:
I am not sure if I understand you correctly but the flat weight is divided if there are multiple nodes requesting from the same IP address. This protects against Sybil attacks using many node keys.